### PR TITLE
VL53L1X time-of-flight sensor.

### PR DESCRIPTION
**Pull request by markCwatson**
The driver utilizes ST Microelectronics library (which exists in modules\hal\st\sensor\vl53l1x). Platform specific headers and source files used by the library are included and adapted for Zephyr.

The driver can be configured in proj.conf to use an interrupt/polling methods and use of the XSHUT pin on the VL53L1X. All uses were tested successfully.

A sample application was developed to test the driver. Due to Zephyr's desire to limit the number of sample applications for specific drivers, it was decided to not include the sample app in this PR. There is a similar app for another time-of-flight sensor (the VL53L0X). It was recommended to adapt this other sample app to be configurable to use either the VL53L0X or the VL53L1X.

Signed-off-by: Mark Watson <mwatson@prosaris.ca>

### CODE REVIEW

**Comment by markCwatson** [comment]
The compliance checks or failing rn. The STM library has some typedefs that I cannot remove. Also, some of their function names are too long, but again, I can't change them. Finally, there are line length complaints related to my device structure setup for use by `DT_INST_FOREACH_STATUS_OKAY` which I don't think I can do anything about. These are causing the checks to fail

**Comment by markCwatson** [comment]
@MaureenHelm @nashif @stephanosio Would it be possible to receive CI overrides for the typedefs issue mentioned previously?? I am hoping to get this into RC1 by the 27th.

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (xyzzy42) [inline drivers/sensor/vl53l1x/vl53l1.c:436]: `PULL_UP` seems strange here.  Usually PULL would only be specified when the gpio is in input.  Otherwise one would drive the gpio high or low and a passive pull up wouldn't make sense.  Unless you want it to be open drain?

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (xyzzy42) [inline drivers/sensor/vl53l1x/vl53l1.c:429]: You should be able to get the client address from the device tree.  I.e., `reg = <0x29>`.  Putting in the dts but then ignoring it and using a hard-coded value doesn't seem right.

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (xyzzy42) [inline drivers/sensor/vl53l1x/vl53l1.c:325]: Can this happen?  It looks like the driver init does the initialization already.

File: drivers/sensor/vl53l1x/vl53l1_platform.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1_platform.c:178]: Ah yes good point

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:429]: Yes my mistake. Forgot to replace.

**Comment by markCwatson** [comment]
Thank you for the review @xyzzy42. I will address your comments first thing Monday morning (Atlantic Canada time),

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:55]: Ah, oops. Thanks for the catch!

**Comment by markCwatson** [comment]
@MaureenHelm @nashif @stephanosio Would it be possible to receive CI overrides for the typedefs issue mentioned previously?? I am hoping to get this into RC1 by the 27th.

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:436]: Hmm, good question. I am testing using [this vl53l1x breakout board ](https://www.digikey.ca/en/products/detail/pimoroni-ltd/PIM373/9565561) which has the XSHUT pin connected to VDD, so I am not actually using the pin. I am going to add a configuration option for the XSHUT pin, similar to what I did for the INT1 interrupt pin. Thanks for the catch!

Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c]: @avisconti OK latest push adds this change and also adds a `return -EIO` on fail per your comment

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (MaureenHelm) [inline drivers/sensor/vl53l1x/vl53l1.c:92]: Indentation here looks strange. Is your editor configured for indent size = 8?

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (MaureenHelm) [inline drivers/sensor/vl53l1x/vl53l1.c:217]: Indentation here looks strange too

File: samples/sensor/vl53l1x/README.rst
Comment (MaureenHelm) [inline samples/sensor/vl53l1x/README.rst:3]: We are trying to reduce the number of driver-specific sensor sample applications. Can you merge this with the vl53l0x sample?

File: samples/sensor/vl53l1x/prj.conf
Comment (MaureenHelm) [inline samples/sensor/vl53l1x/prj.conf:8]: Don't add commented out code. You also don't need `CONFIG_I2C=y` because it gets selected by the driver.

File: samples/sensor/vl53l1x/sample.yaml
Comment (MaureenHelm) [inline samples/sensor/vl53l1x/sample.yaml:6]: Is there any in-tree board that supports vl53l1x?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add this sensor to `tests/drivers/build_all/sensor/`

File: samples/sensor/vl53l1x/sample.yaml
Comment (markCwatson) [reply samples/sensor/vl53l1x/sample.yaml:6]: I don't think so

File: samples/sensor/vl53l1x/README.rst
Comment (markCwatson) [reply samples/sensor/vl53l1x/README.rst:3]: @MaureenHelm Would it be possible to leave my sample as-is for now, then create a new issue to merge the two and assign it to me?

**Comment by markCwatson** [comment]
@MaureenHelm Please note that I have addressed your comments. Thank you for the review! Would it be possible to receive CI overrides for the typedefs issue mentioned previously?? I am hoping to get this into RC1 by the 27th. CC  @nashif @stephanosio

File: samples/sensor/vl53l1x/README.rst
Comment (MaureenHelm) [reply samples/sensor/vl53l1x/README.rst:3]: No, I don't want to introduce a new sample just to remove it soon after. I would prefer that you remove the sample from this PR and come back with a follow-up PR that updates the sample to support both variants.

File: samples/sensor/vl53l1x/sample.yaml
Comment (MaureenHelm) [reply samples/sensor/vl53l1x/sample.yaml:6]: If there are no in-tree boards with:
```yaml
supported:
  - vl53l1x
```
then this sample will never build in CI

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> Would it be possible to receive CI overrides for the typedefs issue mentioned previously?

I'm ok with making an exception for these typedefs and overriding the compliance check. However, there are twister failures that need to be fixed before this PR can be merged.

**Comment by markCwatson** [comment]
> > Would it be possible to receive CI overrides for the typedefs issue mentioned previously?
> 
> I'm ok with making an exception for these typedefs and overriding the compliance check. However, there are twister failures that need to be fixed before this PR can be merged.

In my PR description, I mentioned the STM module/library is missing an `#include <string>` causing warnings during the build. These warnings are treated as errors by Twister. I'm only seeing the failing tests now after adding the vl53l1x test to `tests/drivers/build_all/sensor/i2c.dtsi`. Any suggestions for how to proceed? I canot edit the STM module to add the missing `#include`.

**Comment by erwango** [comment]
> In my PR description, I mentioned the STM module/library is missing an #include <string.h> 

You can add it in `drivers/sensor/vl53l1x/vl53l1_platform.h`, doing platform adaptation is the purpose of this file.

**Comment by markCwatson** [comment]
> > In my PR description, I mentioned the STM module/library is missing an #include <string.h>
> 
> You can add it in `drivers/sensor/vl53l1x/vl53l1_platform.h`, doing platform adaptation is the purpose of this file.

@erwango Thanks so much for the suggestion. I thought the same thing and I thought I tried adding it before and found it didn't resolve the warnings. However, adding it now has fixed it. Thanks again.

@MaureenHelm I removed the sample application and resolved the build/test errors. Thanks so much for your review and responsiveness. Hopefully we can get this approved and merged. Cheers.

**Review by MaureenHelm — COMMENTED** [review]
The commit message needs to be updated to remove:
```
The sample application reads and modifies the default mode
and Region of Interest (ROI), then fetches and reads data
from the sensor. The sample applicaiton uses and was tested
on the nrf52dk by Nordic.

Note: there are warnings during build due to missing
'#include <string.h>' in the STM library. Might require
a separate issue.
```

**Comment by markCwatson** [comment]
All set @MaureenHelm the commit message has been updated.

CC for the other approval @nashif @stephanosio @erwango

**Review by MaureenHelm — DISMISSED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (erwango) [inline tests/drivers/build_all/sensor/i2c.dtsi:410]: Please put device at the end of the list with a unique reg address.
I guess this will produce a warning as 29 is already used.

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (markCwatson) [reply tests/drivers/build_all/sensor/i2c.dtsi:410]: yes, sorry, I didn't notice how the devices were structured like that. I defaulted to sticking this TOF sensor next to the only other one.

@erwango  Note that at the end of this file, there is a device (`test_i2c_as5600`) with a reg that doesn't follow the convention and has a duplicate reg value of 54. It's probably a typo and should be 65. Should I change it here in my PR and stick my device below with reg 65? See below (before and proposed change).

```
...
test_i2c_icp10125: icp10125@63 {
	compatible = "invensense,icp10125";
	reg = <0x63>;
	temperature-measurement-mode = "normal";
	pressure-measurement-mode = "normal";
};

test_i2c_as5600: as5600@54 {
	compatible = "ams,as5600";
	reg = <0x54>;
};
EOF
```
changed to 
```
test_i2c_icp10125: icp10125@63 {
	compatible = "invensense,icp10125";
	reg = <0x63>;
	temperature-measurement-mode = "normal";
	pressure-measurement-mode = "normal";
};

test_i2c_as5600: as5600@64 {
	compatible = "ams,as5600";
	reg = <0x64>;
};

test_i2c_vl53l1x: vl53l1x@65 {
	compatible = "st,vl53l1x";
	reg = <0x65>;
	int-gpios = <&test_gpio 0 0>;
	xshut-gpios = <&test_gpio 0 0>;
};
EOF
```

Comment (markCwatson) [reply tests/drivers/build_all/sensor/i2c.dtsi:410]: This is what I'll do and can revert if necessary

**Comment by markCwatson** [comment]
Thanks again everyone. @MaureenHelm I made some changes and will require your approval again.

@erwango I made the changes you requested. In addition, I fixed my indentation issues. I didn't have my tab spacing set to 8 locally. I am now using `EditorConfig for VS Code` extension with the `.editorconfig` settings, so now what I see matches what you/github sees.

CC @nashif @stephanosio and @avisconti your request for change is still active (I think).

**Review by MaureenHelm — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (xyzzy42) [inline drivers/sensor/vl53l1x/vl53l1.c:368]: It would nice to the user to pass any errors from vl53l1x_read_sensor() on here.  vl53l1x_channel_get() should return a failure code instead of returning success when there is an error.  vl53l1x_channel_get() would also have to not ignore errors.

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:95]: ok valid point

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:368]: Great catch. Thanks. I went over the return codes too and I feel much better now.

File: drivers/sensor/vl53l1x/vl53l1_platform.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1_platform.c:36]: Doing the following results in failure when `VL53L1_WriteMulti` when called by STM's library (tested on hardware). Replacing all other manual endian conversion with `sys_cpu_to_be16` (and the like) seems OK. I appreciate the suggestion, but I am going to leave this one.

```
VL53L1_Error VL53L1_WriteMulti(VL53L1_Dev_t *pdev, uint16_t reg,
		uint8_t *pdata, uint32_t count)
{
	VL53L1_Error status = VL53L1_ERROR_NONE;
	int32_t status_int = 0;

        reg = sys_cpu_to_be16(reg);

        status_int = i2c_write(pdev->i2c, (unit8_t *)&reg, 2, pdev->i2c_slave_address);
	status_int += i2c_write(pdev->i2c, pdata, count, pdev->i2c_slave_address);

	if (status_int < 0) {
		status = VL53L1_ERROR_CONTROL_INTERFACE;
		LOG_ERR("Failed to write");
	}

	return status;
}
```

**Comment by markCwatson** [comment]
Thanks again everyone. @MaureenHelm and @erwango I made some minor changes and will require your approval again.

@xyzzy42 I made the changes you requested.

@avisconti your request for change is still active and possibly blocking.

CC @nashif @stephanosio

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c]: @avisconti would it be possible for you to remove your request for change?

File: drivers/sensor/vl53l1x/vl53l1_platform.c
Comment (xyzzy42) [reply drivers/sensor/vl53l1x/vl53l1_platform.c:36]: It won't work to use two write calls like that.  At least, it depends on the I2C device.  You'll get a two start, address, data, stop, transactions on I2C, which is usually not considered the same thing.  To do it in one transaction, it would be like i2c_burst_write() that uses i2c_transfer() with two messages.

But there's a note that not every i2c controller can do this and combining the data into one buffer is more compatible.  So, maybe you want to keep the bounce buffer like you've got.

But Zephyr still has a function to do the endianness for you.
```
sys_put_be16(reg, buffer);
```

File: drivers/sensor/vl53l1x/vl53l1_platform_user_config.h
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1_platform_user_config.h:41]: shoot i missed fixing this one. That typo by STM

**Review by avisconti — DISMISSED** [review]

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (MaureenHelm) [inline drivers/sensor/vl53l1x/vl53l1.c:458]: This should be `const`

File: drivers/sensor/vl53l1x/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/vl53l1x/Kconfig:27]: This line is not necessary

File: drivers/sensor/vl53l1x/vl53l1_platform.c
Comment (MaureenHelm) [inline drivers/sensor/vl53l1x/vl53l1_platform.c:54]: Is there a reason you didn't use the `_dt` variants for I2C transactions?

**Review by MaureenHelm — COMMENTED** [review]
Needs rebase

**Comment by markCwatson** [comment]
@MaureenHelm I addressed your recent comments. I did the rebase (and lost my one approval from @avisconti). I converted my i2c usage to the `_dt` interface. Everything was build using my previous sample app and flashed to my nrf_52dk board and tested.

**Review by MaureenHelm — DISMISSED** [review]

File: drivers/sensor/vl53l1x/vl53l1.c
Comment (yperess) [inline drivers/sensor/vl53l1x/vl53l1.c:53]: I can't seem to find where this is implemented, can you point me in the right direction?

Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1.c:53]: Hi @yperess. Thanks for taking a look. This is part of a library by ST Microelectronics which was converted to a Zephyr module by some people before my time. It is related to how Zephyr uses external modules using multiple repositories and west as the manager of all ofthat. 

It exists in `modules\hal\st\sensor\vl53l1x` when imported (using the west manifest) when this driver is included in the build of an application (see [here](https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/vl53l1x/)). You can find the definition of that particular function [here](https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/vl53l1x/api/core/src/vl53l1_api.c#:~:text=VL53L1_Error%20VL53L1_GetRangingMeasurementData).

It is imported by Zephyr in the west.yml [here](https://github.com/markCwatson/zephyr/blob/main/west.yml#:~:text=path%3A%20modules/hal/st).

I hope all of this makes sense.

@stephanosio @erwango @MaureenHelm @nashif Correct me if my description is wrong in any way.

**Review by cfriedt — COMMENTED** [review]
LGTM @markCwatson - can you please rebase on the latest `main` and resolve merge conflicts?

File: drivers/sensor/vl53l1x/vl53l1_platform_user_data.h
Comment (cfriedt) [inline drivers/sensor/vl53l1x/vl53l1_platform_user_data.h:24]: Ah, please don't introduce new typedefs. A simple struct pointer should work.

File: drivers/sensor/vl53l1x/vl53l1_platform_user_data.h
Comment (cfriedt) [inline drivers/sensor/vl53l1x/vl53l1_platform_user_data.h:38]: Same here, please just use a struct pointer

Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1_platform_user_data.h:38]: @cfriedt These typedefs are used throughout the STM library being used by this driver thus cannot be changed. There is a similar driver for a different time-of-flight sensor (see drivers/sensor/vl53l0x) that uses a similar library by STM and also has these typedefs, but doesn't appear to be an issue since it is in the main zephyr tree. I spoke to @MaureenHelm about this and she agreed to override the failing compliance checks.

File: drivers/sensor/vl53l1x/vl53l1_platform_user_data.h
Comment (markCwatson) [reply drivers/sensor/vl53l1x/vl53l1_platform_user_data.h:24]: See my other comment, but I should be able to remove this typedef. It's only the `VL53L1_DEV` that is required by the STM library, I believe.

File: drivers/sensor/vl53l1x/vl53l1_platform_user_data.h
Comment (cfriedt) [reply drivers/sensor/vl53l1x/vl53l1_platform_user_data.h:38]: Ah, in that sense, ok

**Review by cfriedt — DISMISSED** [review]

**Comment by cfriedt** [comment]
@markCwatson - you will still need to rebase on main and resolve conflicts manually. I would be happy to re-approve after that.

**Comment by markCwatson** [comment]
> @markCwatson - you will still need to rebase on main and resolve conflicts manually. I would be happy to re-approve after that.

**Comment by markCwatson** [comment]
Ah shoot, I always do that on GitHub. I closed the PR by mistake when trying to cancel a comment by hitting "Close with comment". Sorry about that.

**Comment by markCwatson** [comment]
Thanks a bunch @cfriedt. @MaureenHelm @erwango @avisconti @nashif @stephanosio, sorry to tag you all again, this branch had a conflict that needed to be resolved, so I had to address it. Would love to get this approved/merged so I don't have to keep doing this and bug everyone to give their approvals again LOL

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (erwango) [inline tests/drivers/build_all/sensor/i2c.dtsi:659]: You deleted the node closure. Missing: `};`

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (markCwatson) [reply tests/drivers/build_all/sensor/i2c.dtsi:660]: @erwango oh, thanks, this was where the conflict was with main, and I must have accidentally deleted it.

**Review by MaureenHelm — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (markCwatson) [reply tests/drivers/build_all/sensor/i2c.dtsi:660]: @erwango I fixed this mistake. Would you mind approving? We have two approvals now, but the requested change is holding things up. Thanks for taking a look and for pointing to the exact reason the test was failing!

**Review by cfriedt — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
Arg....another merge conflict, so now this needs another rebase. :(

@yperess this has been a recurring problem since you reworked `tests/drivers/build_all/sensor/i2c.dtsi`. We need to think if there's a better way to organize this test to avoid the churn.

**Review by teburd — APPROVED** [review]

**Comment by markCwatson** [comment]
> Arg....another merge conflict, so now this needs another rebase. :(
> 
> @yperess this has been a recurring problem since you reworked `tests/drivers/build_all/sensor/i2c.dtsi`. We need to think if there's a better way to organize this test to avoid the churn.

Thanks @MaureenHelm. I addressed the merge conflict. I appreciate everyone taking the time to come back again and again to give their approvals. I am going to tag everyone once again and kindly ask for their approvals so we can push this through before the next merge conflict arises: @cfriedt @erwango @teburd @avisconti @nashif @stephanosio

**Review by MaureenHelm — APPROVED** [review]

**Comment by yperess** [comment]
> Arg....another merge conflict, so now this needs another rebase. :(
> 
> @yperess this has been a recurring problem since you reworked `tests/drivers/build_all/sensor/i2c.dtsi`. We need to think if there's a better way to organize this test to avoid the churn.

I'll take a look and run some ideas by you. I'll make an issue for it so we can discuss options.

**Pull request merged**
